src/rvsdg/egglog_optimizer/loop_strength_red.egg

```diff
@@ -0,0 +1,76 @@
+;; in-loop
+;;
+;; ORIGINAL
+;; a = 0
+;; c = 3
+;; for  i = 0 to n:
+;;     a = i * c
+;;
+;; OPTIMIZED
+;; a = 0
+;; c = 3
+;; d = 0
+;; for i = 0 to n:
+;;     a += d
+;;     d += c
+
+(ruleset loop-strength-reduction)
+
+;; For a      theta, operand  is in the theta and either invariant or constant
+(relation inv (Body Operand))
```

Comment:
We already have an invariant analysis in `loop_invariant.egg` that you can reuse I think.